src/PartitionLoops.cpp

```diff
@@ -0,0 +1,926 @@
+        }
+
+        Expr a = mutate(orig_a), b = mutate(orig_b);
+        Interval ia = bounds_of_expr_in_scope(a, scope);
+        Interval ib = bounds_of_expr_in_scope(b, scope);
+
+        bool depends_on_loop_var =
+            !ia.max.same_as(ia.min) ||
+            !ib.max.same_as(ib.min);
+
+        Expr is_true;
+        Expr is_false;
+
+        if (lt) {
+            is_true = ia.max < ib.min;
+            is_false = ia.min >= ib.max;
+        } else if (le) {
+            is_true = ia.max <= ib.min;
+            is_false = ia.min > ib.max;
+        } else if (gt) {
+            is_true = ia.min > ib.max;
+            is_false = ia.max <= ib.min;
+        } else if (ge) {
+            is_true = ia.min >= ib.max;
+            is_false = ia.max < ib.min;
+        }
+
+        internal_assert(is_true.defined() && is_false.defined());
+
+        if (depends_on_loop_var && is_one(simplify(is_true))) {
+            expr = const_true();
+        } else if (depends_on_loop_var && is_one(simplify(is_false))) {
+            expr = const_false();
+        } else if (a.same_as(orig_a) && b.same_as(orig_b)) {
+            expr = e;
+        } else if (lt) {
+            expr = LT::make(a, b);
+        } else if (le) {
+            expr = LE::make(a, b);
+        } else if (gt) {
+            expr = GT::make(a, b);
+        } else if (ge) {
+            expr = GE::make(a, b);
+        }
+
+    }
+
+    void visit(const LT *op) {
+        visit_cmp(op, op->a, op->b);
+    }
+
+    void visit(const LE *op) {
+        visit_cmp(op, op->a, op->b);
+    }
+
+    void visit(const GT *op) {
+        visit_cmp(op, op->a, op->b);
+    }
+
+    void visit(const GE *op) {
+        visit_cmp(op, op->a, op->b);
+    }
+
+    template<typename LetOrLetStmt, typename StmtOrExpr>
+    void visit_let(const LetOrLetStmt *op, StmtOrExpr &result) {
+        Expr value = mutate(op->value);
+
+        StmtOrExpr body;
+        if (value.type() == Int(32)) {
+            Interval i = bounds_of_expr_in_scope(op->value, scope);
+            scope.push(op->name, i);
+            body = mutate(op->body);
+            scope.pop(op->name);
+        } else {
+            body = mutate(op->body);
+        }
+
+        if (value.same_as(op->value) && body.same_as(op->body)) {
+            result = op;
+        } else {
+            result = LetOrLetStmt::make(op->name, value, body);
+        }
+    }
+
+    void visit(const Let *op) {
+        visit_let(op, expr);
+    }
+
+    void visit(const LetStmt *op) {
+        visit_let(op, stmt);
+    }
+
+public:
+    BoundsBasedSimplify(const string &loop_var, Expr min, Expr max) {
+        scope.push(loop_var, Interval(min, max));
+    }
+};
+
+class PartitionLoops : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const For *op) {
+
+        Stmt body = mutate(op->body);
+
+        // Conservatively only apply this optimization at one loop
+        // level. Comment this out to allow full expansion.
```

Comment:
Cool idea - I just tried that out. It does indeed fully simplify the inner loop and make five copies. However, for the boundary conditions performance test things actually got slightly worse. Maybe the expansion in code size? I don't think the boundary condition handling for the y coord costs anything there, because it's just a matter of setting up pointers to the right scanline outside of the inner loop. 

I'll make a note of this in the source. If we hit use cases where we need to handle more than the innermost loop, we can reevaluate.